sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/ApproximateDistinct.java

```diff
@@ -0,0 +1,533 @@
+   *     hllSketch.add(context.element());
+   *     context.output(...);
+   *     }
+   * }
+   * }</pre>
+   *
+   * <p>Then the {@link org.apache.beam.sdk.transforms.DoFn} can be called like this :
+   * <pre>{@code
+   * PCollection<Object> input = ...;
+   * ApproximateDistinctFn myFn = ApproximateDistinctFn.create(new ObjectCoder());
+   * PCollection<OutputT> = input.apply(ParDo.of(new StatefulCardinality(myFn)));
+   * }</pre>
+   *
+   * @param <InputT>      the type of the elements in the input {@link PCollection}
+   */
+  public static class ApproximateDistinctFn<InputT>
+      extends CombineFn<InputT, HyperLogLogPlus, HyperLogLogPlus> {
+
+    private final int p;
+
+    private final int sp;
+
+    private final Coder<InputT> inputCoder;
+
+    private ApproximateDistinctFn(int p, int sp, Coder<InputT> coder) {
+      this.p = p;
+      this.sp = sp;
+      inputCoder = coder;
+    }
+
+    /**
+     * Returns an {@link ApproximateDistinctFn} combiner with the given input coder.
+     *
+     * @param <InputT>    the type of the elements in the input {@link PCollection}
+     * @param coder       the coder that encodes the elements' type
+     */
+    public static <InputT> ApproximateDistinctFn<InputT> create(Coder<InputT> coder) {
+      try {
+        coder.verifyDeterministic();
+      } catch (Coder.NonDeterministicException e) {
+        throw new IllegalArgumentException("Coder is not deterministic ! " + e.getMessage(), e);
+      }
+      return new ApproximateDistinctFn<>(12, 0, coder);
+    }
+
+    /**
+     * Returns a new {@link ApproximateDistinctFn} combiner with a new precision p but
+     * the other parameters remain unchanged.
+     *
+     * @param p           the precision value for the normal representation
+     */
+    public ApproximateDistinctFn<InputT> withPrecision(int p) {
+      if (p < 4) {
+        throw new IllegalArgumentException(String.format("Expected : p >= 4. Actual : p = %s", p));
+      }
+      return new ApproximateDistinctFn<>(p, this.sp, this.inputCoder);
+    }
+
+    /**
+     * Returns a new {@link ApproximateDistinctFn} combiner with a sparse representation
+     * of precision sp but the other parameters remain unchanged.
+     *
+     * @param sp          the precision of HyperLogLog+' sparse representation
+     */
+    public ApproximateDistinctFn<InputT> withSparseRepresentation(int sp) {
+      if ((sp < this.p || sp > 32) && (sp != 0)) {
+          throw new IllegalArgumentException(String.format("Expected : p <= sp <= 32."
+                  + "Actual : p = %s, sp = %s", this.p, sp));
+      }
+      return new ApproximateDistinctFn<>(this.p, sp, this.inputCoder);
+    }
+
+    @Override
+    public HyperLogLogPlus createAccumulator() {
+      return new HyperLogLogPlus(p, sp);
+    }
+
+    @Override
+    public HyperLogLogPlus addInput(HyperLogLogPlus acc, InputT record) {
+      acc.offer(getBytes(record));
```

Comment:
Use CoderUtils.encode (or something like that - there's a method that encodes the given value to bytes)